lambda/shared/src/main/scala/feral/lambda/tracing/TracedLambda.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.tracing
+
+import cats._
+import cats.data._
+import cats.effect.std.Random
+import cats.effect.{Trace => _, _}
+import natchez._
+import feral.lambda._
+import natchez.noop.NoopSpan
+import natchez.xray.{XRay, XRayEnvironment}
+
+abstract class TracedLambda[F[_]: MonadCancelThrow, G[_], Event] {
```

Comment:
I think we should avoid the inheritance, I feel burned after over-using it in the old `Lambda` encoding 😆  I really like the middleware concept `Lambda[G, Event, Result] => Lambda[F, Event, Result]`.

Reply:
Makes sense. It's now `(EntryPoint[F], F ~> G, Kleisli[F, (Event, Context[F]), Kernel]) => Lambda[G, Event, Result] => Lambda[F, Event, Result]` given an implicit `LiftTrace[F, G]`.

Reply:
Solid. This hints to me that maybe `LiftTrace[F, G]` should provide the `F ~> G`.

Reply:
Although now I am dubious how it works with `TracedIO`, maybe that idea was not so good 🤔